File: Hard/question_3.rb
# --------------ASSOCIATION--------------

# Definition: Association represents a bi-directional relationship between two classes. It's a more generic term and 
# doesn't imply a strong ownership or lifecycle dependency between the objects.

# Code Example:
class Author
    attr_accessor :name

    def initialize(name)
        @name = name
    end
end

class Book
    attr_accessor :title, :author

    def initialize(title, author)
        @title = title
        @author = author
    end

author = Author.new("Robert Martin")
book = Book.new("Clean Code", author)
puts book.author.name


# --------------COMPOSITION--------------

# Definition: Composition implies a strong relationship between two classes where one class contains the other, and the 
# contained class has no independent existence. If the container is destroyed, the contained class is also destroyed.

# Code Example:
class Engine
    def start
        puts "Engine started"
    end
end

class Car
    attr_accessor :make, :engine

    def initialize(make)
        @make = make
        @engine = Engine.new
    end
    def start
        puts "Starting car"
        engine.start
    end
end

car = Car.new("Honda")
car.start
end

File: Hard/question_3.kt
// Association:
// Association is a relationship between two classes where one class refers to another,
// but there is no ownership or containment involved.

// Class representing a Person
class Person(val name: String)

// Class representing a Car associated with a Person
class Car(val owner: Person) {
    // Method to start the engine with a message mentioning the owner's name
    fun startEngine() {
        println("Engine started by ${owner.name}")
    }
}

fun main() {
    // Creating a Person named John
    val person = Person("John")

    // Creating a Car associated with the Person
    val car = Car(person)

    // Starting the car's engine
    car.startEngine()
}

// Composition:
// Composition is a stronger form of association where one class contains another class, and the contained 
// class cannot exist independently of the container class. It implies a strong ownership relationship.

// Class representing an Engine
class Engine {
    // Method to start the engine
    fun start() {
        println("Engine started")
    }
}

// Class representing a Car composed of an Engine
class Car {
    // Private property for the Engine
    private val engine: Engine = Engine()

    // Method to start the car's engine
    fun startEngine() {
        engine.start()
    }
}

fun main() {
    // Creating a Car (now the Engine is created within the Car)
    val car = Car()

    // Starting the car's engine
    car.startEngine()
}

fun main() {
    // Creating an Engine
    val engine = Engine()

    // Creating a Car composed of the Engine
    val car = Car(engine)

    // Starting the car's engine
    car.startEngine()
}
